DS-4065: Functions to stack and unstack text and categorization

chrisfacer opened the pull request:
Needed to support Text - Multi inputs as well as Binary - Grid and Nominal - Multi categorizations in the Unstructured Text widget.

Commit:
Add stacking functions for text analysis

Commit:
Basic unit tests

Commit:
Return original vars if no stacking required

Commit:
Fix check for single var

Commit:
Add UnstackCategorization

Commit:
Proper handling of PickOneMulti

Commit:
Tidy code and expand unit tests

Also added function throwCodeReductionWarning from flipRegression which got missed when I previously migrated the other stacking code.

Commit:
Remove verbs from remotes

Commit:
Add checks for correct question types

Commit:
Improve handling of text vs multi and construct text.label

Commit:
Properly extract common prefix

jrwishart commented on DESCRIPTION:
```diff
@@ -1,7 +1,7 @@
 Type: Package
 Title: Data transformations for flip
-Version: 1.13.0
+Version: 1.15.0
```
Perhaps `14` instead of `15`?

jrwishart commented on R/stackingutilities.R:
```diff
@@ -492,9 +492,241 @@ getMultiOutcomeNames <- function(data)
+#' @importFrom flipFormat TidyLabels ExtractCommonPrefix
+#' @export
+StackTextAndCategorization <- function(text, existing.categorization = NULL, subset = TRUE, weights = NULL) {
+    
```
minor improvement. Might make the code below easier to read with the following definition
`has.existing.categorization <- !is.null(existing.categorization)`

jrwishart replied:
Be easier to read I think if the checks for valid categorization are done early and error early if invalid. The definitions below can also update the other `attr` calls later to `question.type` and make it easier to read.

e.g.
```suggestion
has.existing.categorization <- !is.null(existing.categorization)
single.text.variable <- is.list(text) && length(text) == 1L
question.type <- if (has.existing.categorization) attr(existing.categorization, "questiontype")
if (has.existing.categorization) {
    if (single.text.variable && !question.type %in% c("PickOne", "PickAny"))
        stop("The existing categorization should be a Nominal/Ordinal or Binary - Multi variable set")
    if (!single.text.variable && !question.type %in% c("PickOneMulti", "PickAnyGrid") 
        stop("The existing categorization should be a Nominal/Ordinal - Multi or Binary - Grid variable set")
}
if (single.text.variable) 
     return(list(text = text[[1]],
                    existing.categorization = existing.categorization,
                    subset = subset,
                    weights = weights))
```

jrwishart commented on R/stackingutilities.R:
```diff
@@ -492,9 +492,241 @@ getMultiOutcomeNames <- function(data)
+    fake.codes <- 1:length(labels)
+    names(fake.codes) <- labels
+    fake.codes
+}
```
Instead of using `fakeCodeFrame` could use `setNames` instead?

jrwishart commented on R/stackingutilities.R:
```diff
@@ -492,9 +492,241 @@ getMultiOutcomeNames <- function(data)
+        colnames(text) <- text.names
+        attr(text, "codeframe") <- fakeCodeFrame(text.names)
+        attr(text, "questiontype") <- "TextMulti"
+        question.names <- vapply(text, FUN = function(x) attr(x, "question"), FUN.VALUE = character(1))
```
Instead of defining the anonymous function, can use `vapply` but supply additional args in the final `...` arg.

```suggestion
        question.names <- vapply(text, FUN = attr, FUN.VALUE = character(1L), which = "question")
```

jrwishart commented on tests/testthat/test-stackingutilities.R:
```diff
@@ -8,7 +8,7 @@ test_that("ProcessAndStackDataForRegression", {
     weights <- data.to.stack$weights
     interaction <- data.to.stack$interaction
-    exclude.vars.warning  <- "'None of these' have been removed from the set of predictor variables in 'q5' since they don't appear in the set of outcome variables in 'Brand attitude scores'"
+    exclude.vars.warning  <- "have been removed"
```
The expected warning text was simplified. I would prefer to keep it as is as it can help understand what the unit test is achieving and capturing. If the quotes are given you issue it is probably due to the use of `sQuote`. Can either reconstruct the warning with `paste0` and `sQuote` but that can be tedious.  Alternatively can make a function that throws the warning and just call that with the expected parameters.

Additionally, it is not a part of this PR but I noticed below there is a `suppressWarnings` call on line 23. I would discourage that since it can hide things that manifest later. If there are single warnings I prefer getting an exact match. If multiple expected they can be caught with `testthat::capture_warnings` and then handled with `expect_setequal(caught.warnings, c("warn1", "warn2"))` etc

jrwishart commented on tests/testthat/test-stackingutilities.R:
```diff
@@ -37,3 +37,51 @@ test_that("ProcessAndStackDataForRegression", {
+            NA)
+
+            # Correct dimensions
+            expect_equal(length(stacked$text), 1800)
+            expect_equal(length(stacked$weights), 1800)
+            expect_equal(length(stacked$subset), 1800)
+            expect_equal(length(stacked$inds), 1800)
+            if (existing.case == "binary.grid") {
+                expect_equal(nrow(stacked$existing), 1800)
+            } else if (existing.case == "nominal.multi") {
+                expect_equal(length(stacked$existing), 1800)
+            }
```
The text currently only checks that it runs without error and that the lengths are correct. Not that the contents are stacked correctly. These tests could be improved to validate the internal elements? e.g. something like checking the stacked text output matches the unlisted original multi and the rownames are correctly constructed? and other similar checks for the weights and subset?

jrwishart commented on tests/testthat/test-stackingutilities.R:
```diff
@@ -37,3 +37,51 @@ test_that("ProcessAndStackDataForRegression", {
+
+        })
+    }
+}
```
Could add some other tests to check that the function errors correctly with invalid input (e.g. a single text variable needs a PickOne etc).

jrwishart commented on R/stackingutilities.R:
```diff
@@ -492,9 +492,241 @@ getMultiOutcomeNames <- function(data)
+        # No existing categorization to match against.
+        # Stack the text and the filter and weight
+        st <- stack(lapply(text, as.vector))
+        inds <- paste0(rownames(st), ".", st[, "ind"])
```
`rownames(st)` will be length `n * p` where `n` is the number of unstacked rows and `p` is the number of variables. Should instead the index names be `1:n` and then use `.p` ?
This allows an easier deduction of where the data point came from. 
```suggestion
        inds <- apply(expand.grid(rownames(text), colnames(text)), 1, paste0, collapse = ".")
```
requires more code but will be easier to debug later if there are issues? 
Upon further checking it would then match the structure of `rownames(stacking$data))` on line 612. So I think this change is worth it.

jrwishart replied:
Micro-optimization aside, `st[["ind"]]` is quicker than `st[, "ind"]` when using a data.frame.

jrwishart commented on R/stackingutilities.R:
```diff
@@ -492,9 +492,241 @@ getMultiOutcomeNames <- function(data)
+    if (!any(text.labels %in% categorical.codeframe.labels) && !any(text.labels %in% categorical.secondary.codeframe.labels)) {
+        # No matches for the text labels anywhere
+        # Remove common prefixes
+        text.labels <- flipFormat::TidyLabels(text.labels)
```
Ideally remove the `flipFormat::` and add `@importFrom flipFormat TidyLabels` in the `roxygen`

jrwishart commented on R/stackingutilities.R:
```diff
@@ -492,9 +492,241 @@ getMultiOutcomeNames <- function(data)
+        } else if (attr(existing.categorization, "questiontype") == "PickOneMulti") {
+            # Reorder the variables in the categorization to match those
+            # of the text.
+            m <- match(colnames(existing.categorization), colnames(text))
+            existing.categorization <- existing.categorization[, m]
```
I would suggest adding a unit test where this code is doing its job. Currently the names of the both are the same in the unit test so `m` returns `1:3`. i.e. permute the cols in `existing.categorization` and check they are matched correctly.

jrwishart commented on R/stackingutilities.R:
```diff
@@ -492,9 +492,241 @@ getMultiOutcomeNames <- function(data)
+            m <- match(colnames(existing.categorization), colnames(text))
+            existing.categorization <- existing.categorization[, m]
+            st <- stack(lapply(text, as.vector))
+            inds <- paste0(rownames(st), ".", st[, "ind"])
```
similar update to above, change `inds` so that it is `case_number.variable` instead of `1:N.variable` where `N = n * p`.

jrwishart commented on R/stackingutilities.R:
```diff
@@ -492,9 +492,241 @@ getMultiOutcomeNames <- function(data)
+# variable sets in Displayr where redundant text
+# is often removed from labels.
+#' @importFrom stringi stri_reverse
+removeCommonSuffixText <- function(labels) {
+    reversed.labels <- stri_reverse(labels)
+    common <- longestCommonPrefix(reversed.labels)
+    pattern <- paste0("^", common)
+    reversed.labels <- gsub(pattern, "", reversed.labels)
+    labels <- stri_reverse(reversed.labels)
+}
+
+# From https://stackoverflow.com/questions/28273716/r-implementation-for-finding-the-longest-common-starting-substrings-in-a-set-of
+longestCommonPrefix <- function(x) {
+    # sort the vector
+    x <- sort(x)
+    # split the first and last element by character
+    d_x <- strsplit(x[c(1,length(x))],"")
+    # search for the first not common element and so, get the last matching one
+    der_com <- match(FALSE, do.call("==", d_x)) - 1
+    # if there is no matching element, return an empty vector, else return the common part
+    ifelse(der_com==0 , return(character(0)), return(substr(x[1], 1, der_com)))
+}
```
These functions dont appear to be tested.

jrwishart commented on R/stackingutilities.R:
```diff
@@ -492,9 +492,241 @@ getMultiOutcomeNames <- function(data)
+    # split the first and last element by character
+    d_x <- strsplit(x[c(1,length(x))],"")
+    # search for the first not common element and so, get the last matching one
+    der_com <- match(FALSE, do.call("==", d_x)) - 1
```
This will throw an uninformative warning when used. Suggest changing this function.
```
longestCommonPrefix(c("foobarbaz", "foobarwaz", "foobazzzzz"))
```

jrwishart replied:
```suggestion
longestCommonPrefix <- function(x) {
    # sort the vector
    x <- sort(x)
    # split the first and last element by character
    chars <- strsplit(x[c(1, length(x))], "")
    # Find shortest string length
    n.shortest <- length(chars[[1L]])
    # search for the first not common element and so, get the last matching one
    if (length(chars[[2L]]) != n.shortest)
        chars[[2L]] <- chars[[2L]][1:n.shortest]
    common.prefix.length <- which.max(!chars[[1]] == chars[[2]]) - 1L
    if (common.prefix.length == 0L)
        return(character(0L))
    substr(x[1], 1L, common.prefix.length)
}
```

jrwishart committed:
DS-4065 Update longestCommonPrefix function [revdep skip]

Comparison was being made on vectors of different lengths and warnings
being thrown as a result. Function has been re-written to avoid this

Commit:
More comprehensive unit tests

Commit:
Merge branch 'DS-4065' of https://github.com/Displayr/flipTransformations into DS-4065

Commit:
Tidy and add tests

Update according to JW comments

jrwishart committed:
Merge branch 'master' into DS-4065
